### PR TITLE
Fixing random number generator function parser

### PR DESCRIPTION
**Pull request by pengfej**
A bug reported from Aspect community about random number generator.  The simple fix is we need to set optimized=false, because rand will generate different number every time.

### CODE REVIEW

File: source/base/mu_parser_internal.cc [hunk 1]
```diff
@@ -308,8 +308,10 @@ namespace internal
           parser.DefineFun("log", mu_log, true);
           parser.DefineFun("pow", mu_pow, true);
           parser.DefineFun("erfc", mu_erfc, true);
-          parser.DefineFun("rand_seed", mu_rand_seed, true);
-          parser.DefineFun("rand", mu_rand, true);
+          // We need to set optimized=false, because rand will generate
+          // different number every time.
```
Comment (tjhei) [inline source/base/mu_parser_internal.cc:312]: Write this instead:
Disable optimizations (by passing false) that assume the functions will always return the same value:

**Review by tjhei — APPROVED** [review]
Thanks. I think you forgot to include the new test and output file.

File: source/base/mu_parser_internal.cc [hunk 2]
```diff
@@ -308,8 +308,8 @@ namespace internal
           parser.DefineFun("log", mu_log, true);
           parser.DefineFun("pow", mu_pow, true);
           parser.DefineFun("erfc", mu_erfc, true);
-          // We need to set optimized=false, because rand will generate
-          // different number every time.
+          // Disable optimizations (by passign false) that assume the functions
```
Comment (tjhei) [inline source/base/mu_parser_internal.cc:311]:
```suggestion
          // Disable optimizations (by passing false) that assume the functions
```

**Review by tjhei — APPROVED** [review]

File: source/base/mu_parser_internal.cc (hunk 1 above)
Comment (tamiko) [reply source/base/mu_parser_internal.cc]: Ah!

**Review by tamiko — APPROVED** [review]

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Wow, fascinating! This used to work, as illustrated by @cedrict's run of ASPECT. Do you have any idea what caused the change in behavior?

**Review by masterleinad — APPROVED** [review]

**Comment by tjhei** [comment]
> Wow, fascinating! This used to work.

I am confused about that as well. It seems to be broken in 9.4 or newer but working in 9.3 even though we always passed ``true`` here. There were some refactorings and an update to muparser, that might have broken things:
#13745 #13577 #13581

**Comment by drwells** [comment]
Interesting bug - this has been broken since at least 8.5 since we pass `true` there:

https://github.com/dealii/dealii/blob/dealii-8.5/source/base/function_parser.cc#L261

My best guess is that updating the bundled version 'broke' this.

**Pull request merged**